### PR TITLE
cilium: Add option to masq to source route

### PR DESCRIPTION
**Pull request by borkmann**
See commit msgs.

### CODE REVIEW

**Review by kaworu — APPROVED** [review]
Helm changes LGTM.

File: Documentation/network/concepts/masquerading.rst [hunk 1]
```diff
@@ -166,3 +166,11 @@ option ``egress-masquerade-interfaces: eth0`` can be used.
...
    ``eth+``, all interfaces matching the prefix ``eth`` will be used for
    masquerading.
+
+For the advanced case where the routing layer would select different source addresses
+depending on the destination CIDR, the option ``enable-masquerade-to-route-source: "true"``
+can be used in order to masquerade to the source addresses rather than to the
+primary interface address. The latter is then only considered as a catch-all
+fallback, and for the default routes. For these advanced cases the user needs
+to ensure that there are no overlapping destination CIDRs as routes on the
+relevant masquerading interfaces.
```
Comment (brb) [inline Documentation/network/concepts/masquerading.rst:176]: Nit: It'd be nice to add example (from the commit msg) to illustrate the case.

File: pkg/datapath/iptables/iptables.go [hunk 2]
```diff
@@ -1217,7 +1219,77 @@ func (m *IptablesManager) installMasqueradeRules(prog iptablesInterface, ifName,
...
 	}
 
-	// Masquerade all egress traffic leaving the node
+	// Masquerade egress traffic leaving the node based on source routing
+	//
+	// If this option is enabled, then it takes precedence over the catch-all
+	// MASQUERADE further below.
+	if option.Config.EnableMasqueradeRouteSource {
+		devices := option.Config.GetDevices()
+		if option.Config.EgressMasqueradeInterfaces != "" {
```
Comment (brb) [inline pkg/datapath/iptables/iptables.go]: Maybe we could change the `EgressMasqueradeInterfaces` flag type to `StringSlice`? Then the split below would not be needed.

File: pkg/datapath/iptables/iptables.go [hunk 3]
```diff
@@ -1217,7 +1219,77 @@ func (m *IptablesManager) installMasqueradeRules(prog iptablesInterface, ifName,
...
+			devices = strings.Split(option.Config.EgressMasqueradeInterfaces, ",")
+		}
+		family := netlink.FAMILY_V4
+		if prog == ip6tables {
+			family = netlink.FAMILY_V6
+		}
+		if option.Config.EnableIPv4 && prog == ip4tables ||
```
Comment (brb) [inline pkg/datapath/iptables/iptables.go]: I think we don't need to check for `EnableIPv{4,6}` here, as the callers of the `installMasqueradeRules()` already do the check.

File: pkg/datapath/iptables/iptables.go [hunk 4]
```diff
@@ -1217,7 +1219,77 @@ func (m *IptablesManager) installMasqueradeRules(prog iptablesInterface, ifName,
 		}
 	}
 
-	// Masquerade all egress traffic leaving the node
+	// Masquerade egress traffic leaving the node based on source routing
+	//
+	// If this option is enabled, then it takes precedence over the catch-all
+	// MASQUERADE further below.
+	if option.Config.EnableMasqueradeRouteSource {
+		devices := option.Config.GetDevices()
```
Comment (brb) [inline pkg/datapath/iptables/iptables.go:1227]: Do we actually want to fallback to `--devices` if `--egress-masquerade-interfaces` is not set? Asking, as AFAIK `--devices` are only used to control the NodePort BPF and the BPF masq, while previously  the iptables-based MASQ didn't consider them.

**Review by brb — COMMENTED** [review]
Nice work! A few comments.

**Review by giorio94 — APPROVED** [review]
sig-agent changes LGTM

File: pkg/datapath/iptables/iptables.go (hunk 3 above)
Comment (borkmann) [reply pkg/datapath/iptables/iptables.go]: good point, agree. fixed

File: pkg/datapath/iptables/iptables.go (hunk 2 above)
Comment (borkmann) [reply pkg/datapath/iptables/iptables.go]: Makes sense, in prior cases we just use the plain EgressMasqueradeInterfaces string, but I guess it makes sense to switch to slice and have it the other way around.

File: pkg/datapath/iptables/iptables.go (hunk 4 above)
Comment (borkmann) [reply pkg/datapath/iptables/iptables.go:1227]: The source selection can be tied to devices unless next hop groups are used, so both yes & no with regards to their need. I could bail out with error if r.LinkIndex > 0 and len(devices) is 0.

File: pkg/datapath/iptables/iptables.go (hunk 2 above)
Comment (borkmann) [reply pkg/datapath/iptables/iptables.go]: Ok, done.

Comment (borkmann) [reply pkg/datapath/iptables/iptables.go]: Looks like I'm running into https://github.com/cilium/cilium-cli/issues/1896 where CI breaks given cilium-cli vendors in cilium itself :/

File: pkg/datapath/iptables/iptables.go (hunk 4 above)
Comment (borkmann) [reply pkg/datapath/iptables/iptables.go:1227]: Implemented the above now.

File: pkg/datapath/iptables/iptables.go (hunk 2 above)
Comment (borkmann) [reply pkg/datapath/iptables/iptables.go]: Tobias gave some hints how to work-around this by keeping the old EnableMasqueradeRouteSource alive as string.

**Pull request merged**